### PR TITLE
Feature/ETP-2426: Add Payment Process & Fix Payment Plan Tab

### PR DESCRIPTION
**Pull request by leandroallemandismf**

### CODE REVIEW

File: packages/MainUI/hooks/table/useTableData.tsx [hunk 1]
```diff
@@ -285,7 +285,7 @@ export const useTableData = ({
 
   // Build query
   const query: DatasourceOptions = useMemo(() => {
-    const fieldName = tab.parentColumns[0] || "id";
+    const fieldName = tab.parentColumns[tab.tabLevel] || tab.parentColumns[tab?.parentColumns?.length - 1] || "id";
     const value = parentId;
```
Comment (etendobot) [inline packages/MainUI/hooks/table/useTableData.tsx:289]: **Suggestion**

Add defensive checks to ensure that 'tab.parentColumns' is an array and 'tab.tabLevel' is a valid index before accessing 'tab.parentColumns[tab.tabLevel]'. This avoids potential runtime errors if the structure changes or if 'tab.tabLevel' is out of range.

```javascript
const fieldName = Array.isArray(tab.parentColumns) && typeof tab.tabLevel === 'number' && tab.tabLevel < tab.parentColumns.length ? tab.parentColumns[tab.tabLevel] : tab.parentColumns?.[tab.parentColumns?.length - 1] || "id";
```

**Rationale**: While the updated code provides better handling for multi-level parent columns, directly indexing without array/type checks could cause JavaScript 'undefined' errors in rare scenarios. Defensive programming is considered a best practice in complex UI hooks.

File: packages/MainUI/components/ProcessModal/ProcessDefinitionModal.tsx [hunk 2]
```diff
@@ -458,21 +467,36 @@ function ProcessDefinitionModalContent({ onClose, button, open, onSuccess }: Pro
       </div>
     );
   };
+  console.log({
+    _params: {
```
Comment (etendobot) [inline packages/MainUI/components/ProcessModal/ProcessDefinitionModal.tsx:471]: **Suggestion**

Remove direct console.log statements from production React code. Use Etendo's logger or dedicated UI debug/logging tools gated by environment.

```javascript
// console.log({
//   _params: {
//     ...gridSelection,
//     ...mapKeysWithDefaults(form.getValues()),
//   },
// }); // Remove or use Etendo logger in production
```

**Rationale**: Console.log is discouraged in production code as per Etendo React development guidelines. Accidental leakage of sensitive or excessive debug info is possible. Prefer framework logging or remove for production builds.

File: packages/MainUI/utils/processes/manual/utils.ts [hunk 3]
```diff
@@ -110,3 +119,82 @@ export const getParams = ({
...
+    }
+  }
+
+  recursiveUpdateSelection(result as NestedObject);
+
+  return transformDates(result) as TargetObject;
+}
+
+export function transformDates(obj: TransformableValue): TransformableValue {
+  if (typeof obj === "string") {
```
Comment (etendobot) [inline packages/MainUI/utils/processes/manual/utils.ts:187]: **Suggestion**

Add a JSDoc or example for this utility, highlighting date string format expectations (DD-MM-YYYY to YYYY-MM-DD).

```javascript
export function transformDates(obj: TransformableValue): TransformableValue { // Consider: Add JSDoc or sample usage for advanced recursive date normalization }
```

**Rationale**: Making the string format transformation explicit helps avoid confusion for future contributors and is good practice for shared recursive utilities.

File: packages/MainUI/utils/form/normalizeDates.ts [hunk 4]
```diff
@@ -25,3 +25,20 @@ export function normalizeDates(obj: unknown): unknown {
 
   return obj;
 }
+
+export function transformDates(obj: unknown): unknown {
+  if (typeof obj === "string") {
```
Comment (etendobot) [inline packages/MainUI/utils/form/normalizeDates.ts:30]: **Suggestion**

Consider adding a descriptive JSDoc comment explaining the purpose, parameters, and return value of this utility function.

```javascript
/**
 * Recursively transforms all date strings from 'dd-MM-yyyy' to 'yyyy-MM-dd'.
 * @param obj JSON object or array
 * @returns Transformed object with dates in ISO format
 */
export function transformDates(obj: unknown): unknown {
```

**Rationale**: Etendo and Openbravo guidelines encourage documenting utility functions for maintainability and developer clarity, especially for recursive transformers.

File: packages/MainUI/utils/form/normalizeDates.ts [hunk 5]
```diff
@@ -25,3 +25,20 @@ export function normalizeDates(obj: unknown): unknown {
...
+    const dateRegex = /^(\d{2})-(\d{2})-(\d{4})$/;
+    return dateRegex.test(obj) ? obj.replace(dateRegex, "$3-$2-$1") : obj;
+  }
+
+  if (Array.isArray(obj)) {
+    return obj.map(transformDates);
+  }
+
+  if (obj && typeof obj === "object") {
+    return Object.fromEntries(Object.entries(obj).map(([key, value]) => [key, transformDates(value)]));
```
Comment (etendobot) [inline packages/MainUI/utils/form/normalizeDates.ts:40]: **Suggestion**

Use an explicit null check (`obj !== null`) for object type testing. In JavaScript, null is of type 'object', so this avoids unintended behavior.

```javascript
if (obj !== null && typeof obj === "object") {
```

**Rationale**: Explicit null checks are recommended when testing for objects to prevent walking into null, which is technically an object but cannot be traversed (would cause runtime error).

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

File: packages/MainUI/components/ProcessModal/WindowReferenceGrid.tsx [hunk 6]
```diff
@@ -226,10 +227,30 @@ function WindowReferenceGrid({
     skip: shouldSkipFetch,
   });
 
+  useEffect(() => {
+    if (!records) return;
+
+    onSelectionChange((prev: GridSelectionStructure) => ({
+      ...prev,
```
Comment (etendobot) [inline packages/MainUI/components/ProcessModal/WindowReferenceGrid.tsx:234]: **Suggestion**

If retaining previous _allRows is meaningful when records are null, consider defaulting to previous value instead of always setting to an empty array. However, this is context-specific; your current implementation is correct if a full reset is always needed.

```javascript
onSelectionChange((prev: GridSelectionStructure) => ({
      ...prev,
      [parameter.dBColumnName]: {
        _selection: [],
        _allRows: prev?.[parameter.dBColumnName]?._allRows ?? [],
      },
    }));
```

**Rationale**: Etendo UI best practices point out that selection and row-tracking state must be synchronized, but there may be situations where you want to keep previous selection context if no new records are supplied. This is a minor suggestion for future flexibility.

File: packages/MainUI/hooks/table/useTableData.tsx (hunk 1 above)
Comment (etendobot) [inline packages/MainUI/hooks/table/useTableData.tsx:289]: > [!WARNING]
**⚠️ Blocking Issue**

Add a safety check to ensure tab.parentColumns is a valid array and tab.tabLevel is within bounds, to prevent possible runtime errors.

```javascript
const fieldName = Array.isArray(tab.parentColumns) && tab.parentColumns.length
  ? (tab.parentColumns[tab.tabLevel] || tab.parentColumns[tab.parentColumns.length - 1] || "id")
  : "id";
```

**Rationale**: Without verifying that tab.parentColumns is a proper array and tab.tabLevel is a valid index, there is a risk of accessing properties on undefined/null or running into TypeErrors. Etendo UI code should always defend against unexpected structure from backend metadata.

File: packages/MainUI/components/ProcessModal/ProcessDefinitionModal.tsx [hunk 7]
```diff
@@ -90,14 +99,17 @@ function ProcessDefinitionModalContent({ onClose, button, open, onSuccess }: Pro
   const [result, setResult] = useState<ExecuteProcessResult | null>(null);
   const [isPending, startTransition] = useTransition();
   const [loading, setLoading] = useState(true);
-  const [gridSelection, setGridSelection] = useState<unknown[]>([]);
+  const [gridSelection, setGridSelection] = useState<GridSelectionStructure>({});
```
Comment (etendobot) [inline packages/MainUI/components/ProcessModal/ProcessDefinitionModal.tsx:103]: **Suggestion**

Consider adding a clarifying comment to document the per-entity nature of GridSelectionStructure for maintainers, as Etendo pick-and-execute and multi-grid parameter processes can use more than one grid in the future.

```javascript
const [gridSelection, setGridSelection] = useState<GridSelectionStructure>({}); // Structure per-entity
```

**Rationale**: Helps future maintainers understand grid selection supports multiple entities; aligns with Etendo guidance for multi-entity selection in pick-and-execute processes.

File: packages/MainUI/utils/processes/manual/utils.ts [hunk 8]
```diff
@@ -110,3 +119,85 @@ export const getParams = ({
 
   return params;
 };
+
+export function mapKeysWithDefaults(source: SourceObject): TargetObject {
+  const keyMap: KeyMapConfig = {
```
Comment (etendobot) [inline packages/MainUI/utils/processes/manual/utils.ts:124]: **Suggestion**

Consider adding in-line documentation/comments describing the intent and input/output shapes for this function. That will assist future maintainers.

```javascript
export function mapKeysWithDefaults(source: SourceObject): TargetObject {
```

**Rationale**: Explicit function documentation enhances code clarity, maintainability, and onboarding for complex utility methods.

File: packages/MainUI/utils/processes/manual/utils.ts [hunk 9]
```diff
@@ -110,3 +119,85 @@ export const getParams = ({
...
+    }
+  }
+
+  recursiveUpdateSelection(result as NestedObject);
+
+  return transformDates(result) as TargetObject;
+}
+
+export function transformDates(obj: TransformableValue): TransformableValue {
+  if (typeof obj === "string") {
```
Comment (etendobot) [inline packages/MainUI/utils/processes/manual/utils.ts:190]: **Suggestion**

You may want to document in a short comment what date format(s) are handled ("DD-MM-YYYY" to "YYYY-MM-DD") and if other formats are silently ignored.

```javascript
export function transformDates(obj: TransformableValue): TransformableValue {
```

**Rationale**: Function-level date format documentation ensures consistent usage and easier debugging across the codebase.

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

File: packages/MainUI/utils/processes/manual/utils.ts (hunk 8 above)
Comment (etendobot) [inline packages/MainUI/utils/processes/manual/utils.ts:124]: **Suggestion**

Consider adding a brief JSDoc comment summarizing the main purpose and behavior of this function, as it handles complex mapping and object transformations.

```typescript
/**
 * Maps and transforms the keys in a source payment object to target format, applying defaults per business rules and recursively updating nested selections and date formats.
 * @param source The source payment object
 * @returns The mapped, transformed target object
 */
export function mapKeysWithDefaults(source: SourceObject): TargetObject {
```

**Rationale**: Explicit documentation improves future maintainability and helps other developers quickly understand the function's mapping logic and its implications on payment processing in Etendo.

File: packages/MainUI/utils/form/normalizeDates.ts [hunk 10]
```diff
@@ -25,3 +25,20 @@ export function normalizeDates(obj: unknown): unknown {
 
   return obj;
 }
+
+export function transformDates(obj: unknown): unknown {
+  if (typeof obj === "string") {
+    const dateRegex = /^(\d{2})-(\d{2})-(\d{4})$/;
+    return dateRegex.test(obj) ? obj.replace(dateRegex, "$3-$2-$1") : obj;
```
Comment (etendobot) [inline packages/MainUI/utils/form/normalizeDates.ts:32]: **Suggestion**

Consider adding stricter validation for DD and MM ranges (days: 01-31, months: 01-12), as the current regex allows invalid values such as 99-99-2020.

```javascript
const dateRegex = /^(0[1-9]|[12][0-9]|3[01])-(0[1-9]|1[0-2])-\d{4}$/;
```

**Rationale**: While not required, stricter regex can help avoid unintentionally transforming invalid dates. See Etendo and general JavaScript/TypeScript best practices for validation of date input formats.

File: packages/api-client/src/api/datasource.ts [hunk 11]
```diff
@@ -64,14 +64,18 @@ export class Datasource {
   }
 
   private buildParams(options: DatasourceParams) {
-    const params: Record<string, any> = {
+    const params: Record<string, unknown> = {
       _noCount: "true",
```
Comment (etendobot) [inline packages/api-client/src/api/datasource.ts:68]: **Suggestion**

The switch from Record<string, any> to Record<string, unknown> is a good practice that improves type safety in TypeScript.

```typescript
// No immediate suggestion; usage of Record<string, unknown> is correct for Etendo TypeScript API client context.
```

**Rationale**: Etendo's modern TypeScript modules favor 'unknown' over 'any' for stricter type safety and to encourage explicit validation and handling. This matches current industry and Etendo guidelines.

**Review by etendobot — APPROVED** [review]

File: packages/MainUI/hooks/datasource/useTableDirDatasource.ts [hunk 12]
```diff
@@ -97,11 +97,36 @@ export const useTableDirDatasource = ({ field, pageSize = 75, initialPageSize =
     [tab.fields]
   );
```
Comment (etendobot) [inline packages/MainUI/hooks/datasource/useTableDirDatasource.ts:99]: **Suggestion**

Since URLSearchParams expects a Record<string, string>, consider validating that the mutation by isAddPayment keeps all payload values as strings.

```typescript
baseBody = isAddPayment(baseBody); // Consider validation to ensure the result type fits URLSearchParams input.
```

**Rationale**: URLSearchParams ignores non-string values and may create subtle bugs if a boolean or number is added. Use explicit conversion if needed.

File: packages/MainUI/components/ProcessModal/ProcessDefinitionModal.tsx [hunk 13]
```diff
@@ -54,13 +53,23 @@ import ProcessParameterSelector from "./selectors/ProcessParameterSelector";
...
+import { mapKeysWithDefaults } from "@/utils/processes/manual/utils";
 
 /** Fallback object for record values when no record context exists */
 export const FALLBACK_RESULT = {};
 
 /** Reference ID for window reference field types */
 const WINDOW_REFERENCE_ID = FIELD_REFERENCE_CODES.WINDOW;
 
+export type GridSelectionStructure = {
+  [entityName: string]: {
```
Comment (etendobot) [inline packages/MainUI/components/ProcessModal/ProcessDefinitionModal.tsx:66]: **Suggestion**

No runtime or blocking issues. Typing for grid selection is well-structured and supports future-proof extensibility for multiple grids/tabs. Suggest providing a brief JSDoc comment to guide future maintainers on intended usage.

```javascript
export type GridSelectionStructure = {
  [entityName: string]: {
    _selection: unknown[];
    _allRows: unknown[];
  };
};
```

**Rationale**: Etendo’s frontend conventions encourage explicit type aliases for key data structures, and documentation aids maintainability in multi-entity scenarios.

File: packages/MainUI/components/ProcessModal/ProcessDefinitionModal.tsx [hunk 14]
```diff
@@ -90,14 +99,17 @@ function ProcessDefinitionModalContent({ onClose, button, open, onSuccess }: Pro
...
-  const [gridSelection, setGridSelection] = useState<unknown[]>([]);
+  const [gridSelection, setGridSelection] = useState<GridSelectionStructure>({});
 
   const selectedRecords = graph.getSelectedMultiple(tab);
 
-  const windowReferenceTab = parameters.grid?.window?.tabs?.[0] as Tab;
-  const entityName = windowReferenceTab?.entityName || "";
-  const windowId = tab?.window || "";
+  const firstWindowReferenceParam = useMemo(() => {
+    return Object.values(parameters).find((param) => param.reference === WINDOW_REFERENCE_ID);
```
Comment (etendobot) [inline packages/MainUI/components/ProcessModal/ProcessDefinitionModal.tsx:107]: **Suggestion**

Optional: Consider clarifying with a comment that only the first window reference parameter is supported for populating grid selection and tab data, and if multiple are present, only the first is used in the modal (future-proofing for multi-window scenarios).

```javascript
const firstWindowReferenceParam = useMemo(() => {
  return Object.values(parameters).find((param) => param.reference === WINDOW_REFERENCE_ID);
}, [parameters]);
```

**Rationale**: Etendo process definitions may eventually support more complex scenarios; preemptive comments clarify intent and support.

File: packages/MainUI/components/ProcessModal/ProcessDefinitionModal.tsx [hunk 15]
```diff
@@ -458,21 +466,36 @@ function ProcessDefinitionModalContent({ onClose, button, open, onSuccess }: Pro
       </div>
     );
   };
+  console.log({
+    _params: {
```
Comment (etendobot) [inline packages/MainUI/components/ProcessModal/ProcessDefinitionModal.tsx:470]: **Suggestion**

Remove the 'console.log' debugging statement from production code.

```javascript

```

**Rationale**: Debugging statements like 'console.log' should not be present in production/mainline code according to Etendo UI standards and best practices. Logging in Etendo UI should use the provided logger utilities to ensure consistent, properly controlled output. Leaving console logs in production increases the risk of inadvertently leaking sensitive data to client browsers and degrades maintainability.

File: packages/MainUI/utils/processes/manual/utils.ts (hunk 9 above)
Comment (etendobot) [inline packages/MainUI/utils/processes/manual/utils.ts:190]: **Suggestion**

Consider documenting in a block comment above that this function expects 'DD-MM-YYYY' and rewrites it to 'YYYY-MM-DD' for date normalization.

```javascript
export function transformDates(obj: TransformableValue): TransformableValue {
```

**Rationale**: Etendo coding standards emphasize documentation for data normalization routines and data format expectations.

File: packages/api-client/src/api/datasource.ts [hunk 16]
```diff
@@ -64,14 +64,18 @@ export class Datasource {
...
     };
+    const formatKey = (key: string): string => {
+      const specialKeys = new Set(["ad_org_id", "c_currency_id", "issotrx", "received_from", "c_currency_to_id"]);
 
-    const formatKey = (key: string) => (isWrappedWithAt(key) ? key : `_${key}`);
-    const formatValue = (value: any) => (Array.isArray(value) ? value.join(",") : String(value));
+      return specialKeys.has(key) || isWrappedWithAt(key) ? key : `_${key}`;
+    };
+
+    const formatValue = (value: unknown) => (Array.isArray(value) ? value.join(",") : String(value));
```
Comment (etendobot) [inline packages/api-client/src/api/datasource.ts:79]: **Suggestion**

Consider handling null and undefined explicitly in formatValue to avoid sending 'null' or 'undefined' as strings to the backend.

```javascript
const formatValue = (value: unknown) => (Array.isArray(value) ? value.join(",") : String(value));
```

**Rationale**: Sending 'null' or 'undefined' as strings may result in unwanted behavior in Etendo's backend APIs. Safer to convert nullish values to empty string or omit them.

File: packages/MainUI/utils/form/normalizeDates.ts (hunk 4 above)
Comment (etendobot) [inline packages/MainUI/utils/form/normalizeDates.ts:30]: **Suggestion**

Consider adding unit tests to ensure transformDates reliably handles all edge cases (deeply nested objects, arrays, and non-date strings).

```javascript
export function transformDates(obj: unknown): unknown { /* Consider adding test coverage for DD-MM-YYYY to YYYY-MM-DD transformation, including nested objects and arrays. */
```

**Rationale**: Comprehensive test coverage helps ensure future changes or refactoring do not break date conversion or recursive traversal logic.

File: packages/MainUI/components/ProcessModal/ProcessDefinitionModal.tsx [hunk 17]
```diff
@@ -458,21 +466,36 @@ function ProcessDefinitionModalContent({ onClose, button, open, onSuccess }: Pro
       </div>
     );
   };
```
Comment (etendobot) [inline packages/MainUI/components/ProcessModal/ProcessDefinitionModal.tsx:468]: **Suggestion**

This migration from an array to an object makes gridSelection capable of holding selections for multiple entities. Be sure this behavior matches all usages in modal closure and process reinitialization (e.g., when open changes).

```javascript
setGridSelection({});
```

**Rationale**: While not blocking, initialization changes from '[]' to '{}' can potentially break assumptions where an array is expected. Ensure all usages (including downstream WindowReferenceGrid or any handler) now check for an object rather than an array to avoid runtime errors.

File: packages/MainUI/utils/processes/manual/utils.ts [hunk 18]
```diff
@@ -110,3 +119,85 @@ export const getParams = ({
...
+    "Action Regarding Document": { target: "document_action", default: null },
+    "Converted Amount": { target: "converted_amount", default: null },
+    "Deposit To": { target: "fin_financial_account_id", default: null },
+  };
+
+  const result: TargetObject = {};
+
+  for (const [key, value] of Object.entries(source)) {
+    let mappedValue: PrimitiveValue | NestedObject | SelectionItem[] =
+      value !== "" && value != null ? value : keyMap[key]?.default;
```
Comment (etendobot) [inline packages/MainUI/utils/processes/manual/utils.ts:148]: **Suggestion**

Consider checking explicitly for undefined in addition to empty string and null. Although in most JavaScript data flows, absent values appear as null or empty strings, undefined can occur with keys that haven't been set or when values are explicitly undefined.

```javascript
let mappedValue: PrimitiveValue | NestedObject | SelectionItem[] =
      value !== undefined && value !== "" && value != null ? value : keyMap[key]?.default;
```

**Rationale**: Enhances defensive coding by covering all falsy/unset value permutations, reducing risk of accidental use of undefined values and making the intent clear for Etendo contributors.

File: packages/MainUI/hooks/datasource/useTableDirDatasource.ts [hunk 19]
```diff
@@ -97,11 +97,36 @@ export const useTableDirDatasource = ({ field, pageSize = 75, initialPageSize =
...
 
+  interface BaseBody {
+    [key: string]: unknown;
```
Comment (etendobot) [inline packages/MainUI/hooks/datasource/useTableDirDatasource.ts:101]: **Suggestion**

Expand BaseBody to include the new possible properties introduced through isAddPayment (fin_paymentmethod_id, fin_financial_account_id, issotrx) to maintain accurate typings.

```typescript
interface BaseBody {
  [key: string]: unknown;
  inpfinPaymentmethodId?: string;
  inpissotrx?: string;
  windowId?: string;
  'Deposit To'?: string;
  fin_paymentmethod_id?: string;
  fin_financial_account_id?: string;
  issotrx?: boolean;
}
```

**Rationale**: The isAddPayment function injects new properties (fin_paymentmethod_id, fin_financial_account_id, issotrx: boolean) that are not reflected in BaseBody, which can cause subtle type mismatches. Explicit properties promote type safety and clearer documentation for future maintainers.

File: packages/MainUI/hooks/datasource/useTableDirDatasource.ts [hunk 20]
```diff
@@ -97,11 +97,36 @@ export const useTableDirDatasource = ({ field, pageSize = 75, initialPageSize =
...
+    }
+    const { inpfinPaymentmethodId, inpissotrx, windowId, ...rest } = baseBody;
+    const depositTo = baseBody["Deposit To"];
+
+    const result: BaseBody = {
+      ...rest,
+      ...(inpfinPaymentmethodId && { fin_paymentmethod_id: inpfinPaymentmethodId }),
+      ...(depositTo && { fin_financial_account_id: depositTo }),
+      ...(inpissotrx && { issotrx: inpissotrx === "Y" }),
+    };
```
Comment (etendobot) [inline packages/MainUI/hooks/datasource/useTableDirDatasource.ts:120]: **Suggestion**

To prevent accidental falsy value exclusion (such as when inpissotrx is ''), explicitly check for undefined/null rather than relying on coercive truthiness in merging values.

```typescript
      ...(inpissotrx !== undefined && inpissotrx !== null && { issotrx: inpissotrx === 'Y' }),
```

**Rationale**: The current use of && can inadvertently skip values if 'inpissotrx' is an empty string or falsy. Best practice in Etendo UI data flows is explicit checks for presence to ensure intended business logic is always applied.

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

File: packages/MainUI/components/Toolbar/Toolbar.tsx [hunk 21]
```diff
@@ -143,7 +143,9 @@ const ToolbarCmp: React.FC<ToolbarProps> = ({ windowId, isFormView = false }) =>
   const handleProcessSuccess = useCallback(() => {
     refetchDatasource(tab.id);
     graph.clearSelected(tab);
-  }, [graph, refetchDatasource, tab]);
+    graph.setSelected(tab);
+    refetch();
+  }, [graph, refetch, refetchDatasource, tab]);
```
Comment (etendobot) [inline packages/MainUI/components/Toolbar/Toolbar.tsx:148]: **Suggestion**

If both refetchDatasource and refetch are required, document their different effects, as both triggering may cause redundant loads.

```javascript

```

**Rationale**: Multiple data reloads can cause unnecessary queries/network requests if not coordinated. While not blocking, documentation will help maintainers understand why both are called.

File: packages/MainUI/components/ProcessModal/ProcessDefinitionModal.tsx (hunk 7 above)
Comment (etendobot) [inline packages/MainUI/components/ProcessModal/ProcessDefinitionModal.tsx:103]: **Suggestion**

Adopting a structured GridSelection object provides clearer, type-safe handling if multiple window reference parameters or entities are supported in the future. This aligns with extensibility practices and minimizes accidental UI/logic bugs seen in previous grid-based process modals.

```typescript
// No change required; type-safe initialization makes grid selection handling more robust for multi-parameter/multi-entity processes.
```

**Rationale**: Etendo Classic and MainUI documentation recommend using scalable state structures when handling dynamic process parameters—especially with window reference grids that may support selection for multiple entities. This change anticipates future complexity and matches patterns found in advanced Etendo process modals.

File: packages/api-client/src/utils/search-utils.ts [hunk 22]
```diff
@@ -249,23 +249,44 @@ export class LegacyColumnFilterUtils {
    * Determines if a field is a date/datetime field
    */
   static isDateField(fieldName: string, column: Column): boolean {
-    // Check field name patterns
-    const dateFieldPatterns = ["date", "created", "updated", "time", "timestamp"];
+    // Use field reference codes to accurately determine date fields
+    // Reference codes for date/datetime fields from FIELD_REFERENCE_CODES
```
Comment (etendobot) [inline packages/api-client/src/utils/search-utils.ts:253]: **Suggestion**

Consider documenting (in comments) where the reference codes '15', '16', and '478169542A1747BD942DD70C8B45089C' originate—e.g., the AD_Reference or official Etendo docs—for future maintainability.

```javascript
// Use field reference codes from AD_Reference for exact date type detection
```

**Rationale**: Providing clear documentation for hardcoded reference codes improves maintainability and helps new contributors understand the mapping to Etendo model metadata.

File: packages/api-client/src/utils/search-utils.ts [hunk 23]
```diff
@@ -249,23 +249,44 @@ export class LegacyColumnFilterUtils {
...
+    // Get reference from the correct location in the column structure
+    const columnReference = column.reference || (column as any).column?.reference;
 
-    // Check column type
-    let isDateType = false;
-    if (column.type && typeof column.type === "string") {
-      const lowerType = column.type.toLowerCase();
-      isDateType = lowerType.includes("date") || lowerType.includes("time") || lowerType === "datetime";
+    // Debug logging for documentNo specifically
+    if (fieldName === "documentNo") {
```
Comment (etendobot) [inline packages/api-client/src/utils/search-utils.ts:264]: **Suggestion**

Ensure that debug logging for 'documentNo' is removed or guarded by a debug flag/env check before deploying to production to avoid noisy logs.

```javascript
// Consider using a logger/console.debug with a log level check (or remove in production)
```

**Rationale**: Leaving arbitrary console.log debug statements in production code can clutter logs and may lead to inadvertent information exposure or confuse maintainers.

**Review by etendobot — APPROVED** [review]

File: packages/MainUI/components/ProcessModal/WindowReferenceGrid.tsx [hunk 24]
```diff
@@ -226,10 +227,30 @@ function WindowReferenceGrid({
     skip: shouldSkipFetch,
   });
 
+  useEffect(() => {
+    if (!records) return;
```
Comment (etendobot) [inline packages/MainUI/components/ProcessModal/WindowReferenceGrid.tsx:231]: **Suggestion**

Consider passing the current records to the _allRows property when initializing/resetting selection to improve consistency with downstream logic that may depend on the list of all available rows.

```javascript
useEffect(() => {
    if (!records) return;

    onSelectionChange((prev: GridSelectionStructure) => ({
      ...prev,
      [parameter.dBColumnName]: {
        _selection: [],
        _allRows: records,
      },
    }));
  }, [records, onSelectionChange, parameter.dBColumnName]);
```

**Rationale**: Setting _allRows: records (instead of an empty array) aligns with the GridSelectionStructure expectation seen in Etendo UI grid component implementations, and avoids possible confusion or edge cases where a consumer expects _allRows to reflect the available records. Ensuring consistency on all relevant usages improves overall grid state sync.

File: packages/MainUI/utils/processes/manual/utils.ts (hunk 8 above)
Comment (etendobot) [inline packages/MainUI/utils/processes/manual/utils.ts:124]: **Suggestion**

Consider adding JSDoc comments for major utility functions like mapKeysWithDefaults to clarify their expected input/output and business purpose.

```javascript
export function mapKeysWithDefaults(source: SourceObject): TargetObject {
```

**Rationale**: Documentation improves maintainability and developer onboarding, especially for utility functions handling nontrivial business transformations.

File: packages/api-client/src/utils/search-utils.ts [hunk 25]
```diff
@@ -249,23 +249,26 @@ export class LegacyColumnFilterUtils {
...
+    const columnReference = column.reference || (column as any).column?.reference;
 
-    // Check column type
-    let isDateType = false;
-    if (column.type && typeof column.type === "string") {
-      const lowerType = column.type.toLowerCase();
-      isDateType = lowerType.includes("date") || lowerType.includes("time") || lowerType === "datetime";
+    // Primary check: Use reference codes for accurate date field identification
+    if (columnReference && DATE_REFERENCE_CODES.includes(columnReference)) {
+      return true;
```
Comment (etendobot) [inline packages/api-client/src/utils/search-utils.ts:265]: **Suggestion**

Optionally add a fallback (with a comment) for rare fields that may be dates but lack a reference code, warning that this intentionally limits false positives.

```javascript
if (columnReference && DATE_REFERENCE_CODES.includes(columnReference)) {
  return true;
}
```

**Rationale**: Explicitly noting intentional design limits helps future refactoring and bug-tracing.
